lib/AccessStaticPropertyWithinModelContextRule.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SlamPhpStan;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\StaticPropertyFetch;
+use PHPStan\Analyser\Scope;
+use PHPStan\Broker\Broker;
+use PHPStan\Rules\Rule;
+
+final class AccessStaticPropertyWithinModelContextRule implements Rule
+{
+    /**
+     * @var Broker
+     */
+    private $broker;
+
+    /**
+     * @var string
+     */
+    private $modelBaseClassOrInterface;
+
+    /**
+     * @var string
+     */
+    private $singletonAccessor;
+
+    public function __construct(Broker $broker, string $modelBaseClassOrInterface, string $singletonAccessor)
+    {
+        $this->broker                    = $broker;
+        $this->modelBaseClassOrInterface = $modelBaseClassOrInterface;
+        $this->singletonAccessor         = $singletonAccessor;
+    }
+
+    public function getNodeType(): string
+    {
+        return StaticPropertyFetch::class;
+    }
+
+    /**
+     * @param \PhpParser\Node\Expr\StaticPropertyFetch $node
+     * @param \PHPStan\Analyser\Scope                  $scope
+     *
+     * @return string[] errors
+     */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (! $node->class instanceof Node\Name || ! $node->name instanceof Node\VarLikeIdentifier) {
+            return [];
+        }
+
+        if (! $scope->isInClass()) {
+            return [];
+        }
+
+        $classReflection = $scope->getClassReflection();
+        if (null === $classReflection || ! $classReflection->isSubclassOf($this->modelBaseClassOrInterface)) {
+            return [];
+        }
+
+        $baseYiiClassName = (string) $node->class;
+        if (! $this->broker->hasClass($baseYiiClassName)) {
+            return [];
+        }
+
+        $baseYiiClass = $this->broker->getClass($baseYiiClassName);
+        if ($baseYiiClassName !== $this->singletonAccessor && ! $baseYiiClass->isSubclassOf($this->singletonAccessor)) {
+            return [];
+        }
+
+        return [\sprintf('Class %s extends or implements %s and uses %s::$%s: in a model accessing to a singleton is considered an anti-pattern',
```

Comment:
You should check out whether it actually `extends` or `implements` this so that you can provide a better error message to the user.

Reply:
At this point the inspected class is already checked as a subclass of `$this->modelBaseClassOrInterface` and the used singleton as a subclass of `$this->singletonAccessor`.
The user gets in the message:

1. The class name that uses the singleton
1. The extended or implemented class or interface prohibited by configuration
1. The class name of the singleton prohibited
1. The singleton's variable used

What am I missing to provide in the message?

Reply:
If `$this->modelBaseClassOrInterface` is a class, the message should say:

```
Class %s extends %s and uses...
```

If it's an interface:

```
Class %s implements %s and uses...
```

Reply:
Oh, thanks for you patience :+1: